File: dep/guifast_shared/frontend/src/serialization/module_serde.ts
import { ComponentSerde, RequireInfoSerde, TopMenuSerde } from "guifast_shared";

export interface ModuleSerde {
    readonly components: Array<ComponentSerde>;
    readonly menu: Array<TopMenuSerde> | null;
    readonly name: string;
    readonly main_reducer: RequireInfoSerde | null;
    readonly renderer_reducer: RequireInfoSerde | null;
    readonly startup_windows: Array<string> | null;
}

File: dep/guifast_shared/backend/src/serialization/module_serde.rs
use serialization::{ ComponentInfoSerde, RequireInfoSerde, TopMenuSerde };

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct ModuleSerde {
    pub components: Vec<ComponentInfoSerde>,
    pub menu: Option<Vec<TopMenuSerde>>,
    pub name: String,
    pub main_reducer: Option<RequireInfoSerde>,
    pub renderer_reducer: Option<RequireInfoSerde>,
    pub startup_windows: Option<Vec<String>>
}

impl ModuleSerde {
    pub fn new(
        components: Vec<ComponentInfoSerde>,
        menu: Option<Vec<TopMenuSerde>>,
        name: String,
        main_reducer: Option<RequireInfoSerde>,
        renderer_reducer: Option<RequireInfoSerde>,
        startup_windows: Option<Vec<String>>
    ) -> Self {
        ModuleSerde {
            components: components,
            menu: menu,
            name: name,
            main_reducer: main_reducer,
            renderer_reducer: renderer_reducer,
            startup_windows: startup_windows
        }
    }
}
